### PR TITLE
Implement Scott's method for fast cofactor multiplication.

### PR DESCRIPTION
**Pull request by mmaker**
I've recently been looking at faster way for multiplying by the cofactor over G2.  It seems to me that this problem is generally solved in the literature by writing the cofactor (or a multiple of it) modulo q, and using an endomorphism to speed up the computation for scalar multiplication modulo q.

There's two viable methods for BLS12-381: one from [Scott et al.](https://eprint.iacr.org/2008/530.pdf), one from [Budroni et al.](https://eprint.iacr.org/2017/419.pdf). Together with @AnitaDurr we took a chance at implementing Scott's method. There are some trivial optimizations that can still be put in place, but the bottom line is that we can multiply by the cofactor over G2 with 3 scalar multiplications (< r) and 2 point additions.

It seems that relic has also done something [similar](https://github.com/relic-toolkit/relic/blob/master/src/epx/relic_ep2_map.c#L93), however they keep the constant `x` around and use Alessandro's method, making this slightly slower.

### CODE REVIEW

File: src/bls12_381/ec.rs [hunk 1]
```diff
@@ -1557,20 +1557,70 @@ pub mod g2 {
...
+
+            let wx = u.pow(&[
+                0x9354ffffffffe38e,
+                0x0a395554e5c6aaaa,
+                0xcd104635a790520c,
+                0xcc27c3d6fbd7063f,
+                0x190937e76bc3e447,
+                0x08ab05f8bdd54cde,
+            ]);
+            let wx = wx.inverse().unwrap();
```
Comment (daira) [inline src/bls12_381/ec.rs]: `wx` and `wy` could be precomputed. (I'd suggest still computing them this way, just in advance.)

Comment (daira) [reply src/bls12_381/ec.rs]: Please add a comment explaining where the exponents come from.

File: src/bls12_381/ec.rs [hunk 2]
```diff
@@ -1557,20 +1557,70 @@ pub mod g2 {
...
-            let cofactor = BitIterator::new([
-                0xcf1c38e31c7238e5,
-                0x1616ec6e786f0c70,
-                0x21537e293a6691ae,
-                0xa628f1cb4d9e82ef,
-                0xa68a205b2e5a7ddf,
-                0xcd91de4547085aba,
-                0x91d50792876a202,
-                0x5d543a95414e7f1,
+
```
Comment (daira) [inline src/bls12_381/ec.rs]: Add a comment referencing the paper this comes from, and briefly explaining what's going on.

File: src/bls12_381/ec.rs [hunk 3]
```diff
@@ -1557,20 +1557,70 @@ pub mod g2 {
...
+            ]);
+            let wx = wx.inverse().unwrap();
+
+            self.x.frobenius_map(1);
+            self.x.mul_assign(&wx);
+            self.y.frobenius_map(1);
+            self.y.mul_assign(&wy);
+        }
+
+        fn scale_by_cofactor(&self) -> G2 {
```
Comment (daira) [inline src/bls12_381/ec.rs]: This looks like a less efficient method than Wahby and Boneh (see https://github.com/zcash/zcash/issues/3425#issuecomment-487320929 ). Their method takes only two multiplications by the parameter x.

File: src/bls12_381/ec.rs [hunk 4]
```diff
@@ -2012,6 +2062,33 @@ pub mod g2 {
...
+        }
+
+        let mut rng = XorShiftRng::from_seed([0x5dbe6259, 0x8d313d76, 0x3237db17, 0xe5bc0654]);
+        for _i in 0 .. 20 {
+            let p = random_point(&mut rng);
+            let got = p.scale_by_cofactor().into_affine();
+            assert!(got.is_on_curve());
+            assert!(got.is_in_correct_subgroup_assuming_on_curve());
+        }
+    }
```
Comment (daira) [inline src/bls12_381/ec.rs]: I think we should also test that the new `scale_by_cofactor` matches the old one for random curve points (not necessarily in the subgroup).

**Comment by mmaker** [comment]
Hello @daira! Sorry for the very late reply on this pull request. 
Following the big refactoring that I see on https://github.com/zkcrypto/bls12_381, I'm moving my pull request there, integrating with your comments!

Thanks for this preliminary review.

**Pull request closed without merging**